### ISSUE TITLE
Deadlock when spawning engine with large Image.asset

### ISSUE THREAD

**Issue by gaaclarke** [flutter/flutter#93564]
## Seen result
Deadlock when spawning 10 engines from a FlutterEngineGroup quickly with an module that contains a large image from a package.

## Expected result
No deadlock.

## Reproduction code
https://github.com/gaaclarke/flutter/tree/flutter_assets_deadlock

## Reproduction steps

Launch the project at `dev/benchmarks/multiple_flutters/android`.  The hang will happen when the app is started.

## Notes

If you remove the Image, the hang no longer happens:
```diff
--- a/dev/benchmarks/multiple_flutters/module/lib/main.dart
+++ b/dev/benchmarks/multiple_flutters/module/lib/main.dart
@@ -150,10 +150,6 @@ class _MyHomePageState extends State<MyHomePage> {
                     color: Colors.teal,
                     size: 36.0,
                   ),
-                  Image.asset('food/fruits.png',
-                      package: 'flutter_gallery_assets',
-                      width: 36.0,
-                      height: 36.0),
                 ],
               ),
               CustomPaint(
```

If you look at the console you'll see `D/PathProviderPlugin: Use TaskQueues.` get printed out as each path_provider plugin is registered.  It won't get printed 10 times so it seems to be related to an issue that happens during plugin registration.

This hang could be related to the work done in Lightweight Flutter Engines, and it might be a deadlock that is always possible, it's just that Lightweight Flutter Engines is stressing the code in ways that typically doesn't happen.

## Stacktrace of deadlock

I couldn't get a full C++ stacktrace, I can't figure out how to get the local engine running with the add-to-app setup.

```
"main@19558" prio=5 tid=0x2 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at io.flutter.embedding.engine.FlutterJNI.nativeSpawn(FlutterJNI.java:-1)
	  at io.flutter.embedding.engine.FlutterJNI.spawn(FlutterJNI.java:352)
	  at io.flutter.embedding.engine.FlutterEngine.spawn(FlutterEngine.java:396)
	  at io.flutter.embedding.engine.FlutterEngineGroup.createAndRunEngine(FlutterEngineGroup.java:133)
	  at io.flutter.embedding.engine.FlutterEngineGroup.createAndRunEngine(FlutterEngineGroup.java:99)
	  at dev.flutter.multipleflutters.MainActivity$onCreate$engines$2.invoke(MainActivity.kt:45)
	  at dev.flutter.multipleflutters.MainActivity$onCreate$engines$2.invoke(MainActivity.kt:21)
	  at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:172)
	  at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:722)
	  at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:752)
	  at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:743)
	  at dev.flutter.multipleflutters.MainActivity.onCreate(MainActivity.kt:46)
	  at android.app.Activity.performCreate(Activity.java:8051)
	  at android.app.Activity.performCreate(Activity.java:8031)
	  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
	  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
	  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
	  at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
	  at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	  at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
	  at android.os.Handler.dispatchMessage(Handler.java:106)
	  at android.os.Looper.loopOnce(Looper.java:201)
	  at android.os.Looper.loop(Looper.java:288)
	  at android.app.ActivityThread.main(ActivityThread.java:7838)
	  at java.lang.reflect.Method.invoke(Method.java:-1)
	  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)

"FinalizerWatchdogDaemon@20036" daemon prio=5 tid=0x1d nid=NA waiting
  java.lang.Thread.State: WAITING
	 blocks FinalizerWatchdogDaemon@20036
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:442)
	  at java.lang.Object.wait(Object.java:568)
	  at java.lang.Daemons$FinalizerWatchdogDaemon.sleepUntilNeeded(Daemons.java:341)
	  - locked <0x4e77> (a java.lang.Daemons$FinalizerWatchdogDaemon)
	  at java.lang.Daemons$FinalizerWatchdogDaemon.runInternal(Daemons.java:321)
	  at java.lang.Daemons$Daemon.run(Daemons.java:139)
	  at java.lang.Thread.run(Thread.java:920)

"FinalizerDaemon@20038" daemon prio=5 tid=0x1c nid=NA waiting
  java.lang.Thread.State: WAITING
	 blocks FinalizerDaemon@20038
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:442)
	  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:190)
	  - locked <0x4e78> (a java.lang.Object)
	  at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:211)
	  at java.lang.Daemons$FinalizerDaemon.runInternal(Daemons.java:273)
	  at java.lang.Daemons$Daemon.run(Daemons.java:139)
	  at java.lang.Thread.run(Thread.java:920)

"flutter-worker-0@19612" prio=5 tid=0x26 nid=NA waiting
  java.lang.Thread.State: WAITING
	  at sun.misc.Unsafe.park(Unsafe.java:-1)
	  at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:230)
	  at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:461)
	  at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
	  at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:937)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1091)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	  at java.lang.Thread.run(Thread.java:920)

"AsyncTask #1@19615" prio=5 tid=0x27 nid=NA waiting
  java.lang.Thread.State: WAITING
	  at sun.misc.Unsafe.park(Unsafe.java:-1)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:190)
	  at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:459)
	  at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
	  at java.util.concurrent.SynchronousQueue.take(SynchronousQueue.java:920)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1092)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	  at java.lang.Thread.run(Thread.java:920)

"flutter-worker-1@19618" prio=5 tid=0x28 nid=NA waiting
  java.lang.Thread.State: WAITING
	  at sun.misc.Unsafe.park(Unsafe.java:-1)
	  at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:230)
	  at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:461)
	  at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
	  at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:937)
	  at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1091)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	  at java.lang.Thread.run(Thread.java:920)

"ReferenceQueueDaemon@20039" daemon prio=5 tid=0x1b nid=NA waiting
  java.lang.Thread.State: WAITING
	 blocks ReferenceQueueDaemon@20039
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Object.wait(Object.java:442)
	  at java.lang.Object.wait(Object.java:568)
	  at java.lang.Daemons$ReferenceQueueDaemon.runInternal(Daemons.java:217)
	  - locked <0x3c38> (a java.lang.Class)
	  at java.lang.Daemons$Daemon.run(Daemons.java:139)
	  at java.lang.Thread.run(Thread.java:920)

"HeapTaskDaemon@20037" daemon prio=5 tid=0x1a nid=NA waiting
  java.lang.Thread.State: WAITING
	  at dalvik.system.VMRuntime.runHeapTasks(VMRuntime.java:-1)
	  at java.lang.Daemons$HeapTaskDaemon.runInternal(Daemons.java:531)
	  at java.lang.Daemons$Daemon.run(Daemons.java:139)
	  at java.lang.Thread.run(Thread.java:920)

"Signal Catcher@20034" daemon prio=10 tid=0x18 nid=NA waiting
  java.lang.Thread.State: WAITING

"ADB-JDWP Connection Control Thread@20035" daemon prio=0 tid=0x19 nid=NA waiting
  java.lang.Thread.State: WAITING

"Jit thread pool worker thread 0@20040" daemon prio=5 tid=0x1e nid=NA runnable
  java.lang.Thread.State: RUNNABLE

"Binder:15395_1@20041" prio=5 tid=0x1f nid=NA runnable
  java.lang.Thread.State: RUNNABLE

"Binder:15395_2@20042" prio=5 tid=0x20 nid=NA runnable
  java.lang.Thread.State: RUNNABLE

"Binder:15395_3@20043" prio=5 tid=0x21 nid=NA runnable
  java.lang.Thread.State: RUNNABLE

"Profile Saver@19561" daemon prio=5 tid=0x25 nid=NA runnable
  java.lang.Thread.State: RUNNABLE

"RenderThread@19619" daemon prio=7 tid=0x29 nid=NA runnable
  java.lang.Thread.State: RUNNABLE

"1.ui@20033" prio=6 tid=0x2a nid=NA runnable
  java.lang.Thread.State: RUNNABLE
```

cc @zanderso @jason-simmons

**Comment by gaaclarke** [comment]
@jason-simmons I'm not sure if you can get it, but it would be nice to get a c++ stacktrace for this hang so we can get a sense of the cause of this issue to understand the priority.  There is a risk the issue could affect regular usage of Flutter and this usage is just causing it to manifest more often.  I couldn't figure out how to get it.

**Comment by jason-simmons** [comment]
This deadlock is happening because `Shell::Spawn` is constructing the child shell within a `GetIsGpuDisabledSyncSwitch()->Execute()` task.
(see https://github.com/flutter/engine/blob/main/shell/common/shell.cc#L512)

`SyncSwitch::Execute` holds the SyncSwitch's mutex while executing the function.  The SyncSwitch task executed by `Shell::Spawn` is calling `Shell::CreateWithSnapshot`/`CreateShellOnPlatformThread`.  That is a complex multi-step process that depends on tasks queued to other threads.

In particular, `CreateShellOnPlatformThread` posts a task to the IO thread to create an IO manager for the shell and then waits for its completion.

Before that task runs, there may be another task queued to the IO thread that needs to acquire the parent shell's GPU SyncSwitch mutex.  In this app, that task is an image decoder that is calling `UploadRasterImage`.

So the IO thread is blocked trying to acquire the SyncSwitch mutex.  Meanwhile, the platform thread owns the SyncSwitch mutex but is blocking waiting for an IO thread task to complete.  That results in a deadlock.

@gaaclarke Why does `Shell::Spawn` create the child shell within a `SyncSwitch::Execute`?

**Comment by gaaclarke** [comment]
Thanks Jason, I'd be interested in hearing how you got a C++ stacktrace when you get a chance.

So, it's calling `GetIsGpuDisabledSyncSwitch()->Execute` specifically.  The only reasons to use that is to block the app from backgrounding while the GPU is in use on the io or raster thread, or in this case to look at the switch's value.  There is no accessor for the gpu disable sync switch value.  This forces users to place their sensitive code inside of the lambda functions, but in this case we've put too much logic inside of that lambda function.

Since Shell::Spawn happens on the platform thread, we won't have to worry about the value of the sync switch changing in this event (it is an event on the platform thread that sets it).  I believe the concern was that we'd be asked to spawn an engine, look to see if the gpu is disabled, then the value changes before the engine is actually created.  Since that isn't something that can't happen in practice the code could read:

```c++
  std::unique_ptr<Shell> result;
  bool isGpuDisabled = true;
  GetIsGpuDisabledSyncSwitch()->Execute(
      fml::SyncSwitch::Handlers()
          .SetIfFalse([&] { isGpuDisabled = false; })
          .SetIfTrue([&] { isGpuDisabled = false; }));
  result = shell_maker(isGpuDisabled);
  result->shared_resource_context_ = io_manager_->GetSharedResourceContext();
  result->RunEngine(std::move(run_configuration));
  return result;
```

If we do this this though it could be a problem if we ever have an embedder that tries to set the GpuDisabledSyncSwitch from a thread other than the platform thread.  So I'd also add an assert to `Shell::SetGpuAvailability` as well.
